### ISSUE TITLE
NameSpaceImport behavior is different from NamedImports behavior

### ISSUE THREAD

**Issue by lPadier** [babel/babel#2276]
export.js
```js
export var value = 0;
```

import1.js
```js
import {value} from "./export";
value++; // SyntaxError: import.js: Line 2: "value" is read-only
```

import2.js
```js
import * as export from "./export";
export.value++; // No error
```

If I understand the specification properly, both of these should throw an error as all the properties of export should be immutable bindings

**Comment by sebmck** [comment]
Can you please cite some spec references? How do you propose this be caught?

**Comment by lPadier** [comment]
Part 12.c.iii of http://www.ecma-international.org/ecma-262/6.0/#sec-moduledeclarationinstantiation seems to indicate it. I'm not sure exactly how the first one is called in Babel, but would a similar system be possible for the properties of an object ?

**Comment by loganfsmyth** [comment]
My reading of http://www.ecma-international.org/ecma-262/6.0/#sec-module-namespace-exotic-objects-set-p-v-receiver does make it sound like module namespace exports are not mutable, even when imported this way.

**Comment by loganfsmyth** [comment]
That said, this'll be a pain and would hurt performance since it'd require more runtime code AND it would have to be per-module-type :(

**Comment by lPadier** [comment]
One possible solution would be to change the way exports work when converted to commonJS modules.

```js
export var value = 0;
```
is transformed into:

Before change:
```js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var value = 0;
exports.value = value;
```

After change:
```js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var value = 0;

Object.defineProperty(exports, "value", {
  get() {return value;},
  set() {throw new TypeError();}
}
```

**Comment by lPadier** [comment]
This works for this small example, but I don't think it would work if value is updated inside the module

**Comment by loganfsmyth** [comment]
Yup, that's basically the alternative, but it means every export access will call a function instead of access an object property, which will probably be slower. e.g.

    import {value} from 'mod';

    for (var i = 0; i < 10000; i++) value();

would now call a function 10k times. Seems like this behavior would have to be optional?

**Comment by lPadier** [comment]
This could be a different module formatter.

Will module formatters also be moved into plugins after 6.0.0 or it this just transformers ?

**Comment by jmm** [comment]
@lPadier 
> Will module formatters also be moved into plugins after 6.0.0 or it this just transformers ?

That is my understanding based on discussion on slack (https://github.com/babel/babel/issues/2212#issuecomment-137177477). Would probably be good to add that to the v6 tracking issue.

**Comment by nicolo-ribaudo** [comment]
Would it be enough to rewrite imports to
```js
var imported = Object.freeze(_interopRequireWildcard(require("./file")));
```
?

**Comment by loganfsmyth** [comment]
If we wanted to do it properly, we'd make every export define a getter with no setter instead. Values are allowed to change, they just aren't writable from the outside.
